modules/aggregations/src/main/java/org/elasticsearch/aggregations/bucket/timeseries/TimeSeriesAggregator.java

```diff
@@ -92,15 +92,29 @@ protected void doClose() {
     protected LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCtx, LeafBucketCollector sub) throws IOException {
         return new LeafBucketCollectorBase(sub, null) {
 
+            // This helps significantly reducing time spent attempting to add bucket + tsid combos that already were added.
+            long currentTsidOrd = -1;
+            long currentBucket = -1;
+            long currentBucketOrdinal;
+
             @Override
             public void collect(int doc, long bucket) throws IOException {
+                if (currentBucket == bucket && currentTsidOrd == aggCtx.getTsidOrd()) {
```

Comment:
It looks like this aggregation is always called with `bucket == 0` so maybe remove this `currentBucket == bucket` condition here and replace it with an assertion that `bucket` is zero?

Reply:
I think that is only the case if `time_series` aggregation is a top level aggregation.
I don't think this is always the case and for example a `date_histogram` aggregation is a likely parent aggregation in the case of `time_series` aggregation.

Reply:
Ah, that makes sense. Then maybe the `currentBucket == bucket` condition is a bit fragile as it would prevent the optimization from kicking in if this collector is called with different bucket ordinals even though it the TSID doesn't change. What about changing `currentTsidOrd` and `currentBucketOrdinal` to arrays that are keyed by `bucket`, ie. tracking separately the current TSID and bucket ordinal per bucket?

Reply:
I think the case we know there isn't a parent aggregation then we can return a leaf bucket collector that only does the `currentTsidOrd == aggCtx.getTsidOrd()` check.

Reply:
> What about changing currentTsidOrd and currentBucketOrdinal to arrays that are keyed by bucket, ie. tracking separately the current TSID and bucket ordinal per bucket?

We'd need a BigArray thing because we can have many buckets above us and we'd want to track the memory. It could be useful though. If you ever "come back" to the same bucket. But I don't think, say, `date_histogram` will. I think once it's finished collecting some date for the time series it'll never come back.

>I think the case we know there isn't a parent aggregation then we can return a leaf bucket collector that only does the currentTsidOrd == aggCtx.getTsidOrd() check.

That's what the `CardinalityUpperBound bucketCardinality` argument to the ctor is for. If it's *exactly* one you know you don't have a parent agg *or* it always makes a single bucket.

Reply:
> What about changing currentTsidOrd and currentBucketOrdinal to arrays that are keyed by bucket, ie. tracking separately the current TSID and bucket ordinal per bucket?

I think for `currentTsidOrd` we don't need an array? This is because of how the `TimeSeriesIndexSearcher` executes we only see a tsid once. Once the `TimeSeriesIndexSearcher` has moved the next tsid then we will not see the previous tsid again.

I agree that for bucket ord the optimisation is fragile, only in certain cases we can make use of the caching here.
I think keeping track of current tsid ord in an array can still be useful, but we need to use big arrays like Nik mentioned.

> But I don't think, say, date_histogram will. I think once it's finished collecting some date for the time series it'll never come back.

Right, this is true for `date_histogram` or `auto_date_histogram` on the `@timestamp` field. I have been benchmarking with such an aggregation as parent. This is the ideal case and then keep track of the current bucket order in just a field is enough for the optimisation to kick in.

> That's what the CardinalityUpperBound bucketCardinality argument to the ctor is for. If it's exactly one you know you don't have a parent agg or it always makes a single bucket.

Good point, `bucketCardinality` should then be used to make such a decision.

Reply:
Thanks for the explanation about a parent date_histogram, it makes total sense, collections of (TSID, bucket) tuples are not interleaved given TSDB indexing. Let's add a comment about it to explain why this apparently naive approach actually works very well?

Reply:
I will add a comment. I think the overhead of this caching is minimal now, and let's a terms aggregator is a parent then the fact that a previous bucket ord is emitted isn't too bad, in that case the `BytesKeyedBucketOrds` needs to be checked. This is what happens now all the time.

Reply:
Actually things should be fine with a `terms` aggregation as a parent too, because almost all the time this `terms` aggregation would run on a field that is a dimension or a tag. So `bucket` would only change when `tsidOrd` changes too?

Reply:
Yes, I think that is true if the field of the `terms` agg is a dimension field. But maybe not in the case if the field is a non dimension field (label field)?

Reply:
> My intuition is that we don't have to care much about this case of labels that could take different values for the same TSID (as long as the aggregation output is correct, just less optimized), do we?

This is what I think as well. This change favours parent aggs based on dimension fields, `@timestamp` or no parent aggregation. For the other cases, this change doesn't help, but I think doesn't hurt either. We are always correct, since we then fall back to checking whether we have seen tsid & bucket combination in `BytesKeyedBucketOrds`.

Reply:
I added a comment: f23de4976776c7f7de2b481f59b8f7170ee6630f